File: sdk/toon.ts
// TypeScript implementation of TOON (Token-Oriented Object Notation)
export type JSObject = { [k: string]: any };

function escapeValue(val: any, delimiter = '|'): string {
  let s = val === null || val === undefined ? '' : String(val);
  s = s.replace(/\\/g, '\\\\');
  s = s.replace(/\n/g, '\\n');
  if (delimiter) {
    const re = new RegExp(escapeRegExp(delimiter), 'g');
    s = s.replace(re, '\\' + delimiter);
  }
  return s;
}

function unescapeValue(s: string): string {
  let res = '';
  for (let i = 0; i < s.length; i++) {
    const c = s[i];
    if (c === '\\') {
      i++;
      if (i >= s.length) {
        res += '\\';
        break;
      }
      const nxt = s[i];
      if (nxt === 'n') res += '\n';
      else res += nxt;
    } else {
      res += c;
    }
  }
  return res;
}

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function splitEscaped(line: string, delimiter = '|'): string[] {
  const parts: string[] = [];
  let cur = '';
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (c === '\\') {
      if (i + 1 < line.length) {
        cur += '\\' + line[i + 1];
        i++;
      } else {
        cur += '\\';
      }
    } else if (delimiter && c === delimiter) {
      parts.push(cur);
      cur = '';
    } else {
      cur += c;
    }
  }
  parts.push(cur);
  return parts;
}

export function toToon(objects: JSObject[], fields?: string[], delimiter = '|', types?: string[], includeSchema = false): string {
  if (!objects || objects.length === 0) return '';
  if (!fields) fields = Object.keys(objects[0]);
  const header = fields.join(delimiter);
  let schemaLine = '';
  if (includeSchema) {
    if (!types) {
      types = fields.map((f) => {
        const v = objects[0][f];
        if (v === null || v === undefined) return 'str';
        if (typeof v === 'boolean') return 'bool';
        if (Number.isInteger(v)) return 'int';
        if (typeof v === 'number') return 'float';
        return 'str';
      });
    }
    schemaLine = types.join(delimiter);
  }
  const rows = objects.map((obj) => fields!.map((f) => escapeValue(obj[f], delimiter)).join(delimiter));
  const lines = [header];
  if (includeSchema) lines.push(schemaLine);
  return lines.concat(rows).join('\n');
}

export function fromToon(toon: string, delimiter = '|'): JSObject[] {
  if (!toon) return [];
  const lines = toon.split(/\r?\n/);
  const header = lines[0];
  const fields = header ? header.split(delimiter) : [];
  let types: string[] | undefined;
  let start = 1;
  if (lines.length > 1) {
    const cand = lines[1];
    const candParts = cand.split(delimiter);
    const common = new Set(['int', 'integer', 'float', 'double', 'number', 'str', 'string', 'bool', 'boolean']);
    if (candParts.every((p) => common.has(p.toLowerCase()))) {
      types = candParts;
      start = 2;
    }
  }
  const out: JSObject[] = [];
  for (let i = start; i < lines.length; i++) {
    const parts = splitEscaped(lines[i], delimiter).map((p) => unescapeValue(p));
    while (parts.length < fields.length) parts.push('');
    const obj: JSObject = {};
    for (let j = 0; j < fields.length; j++) {
      const typ = types && types[j] ? types[j] : undefined;
      obj[fields[j]] = castValue(parts[j], typ);
    }
    out.push(obj);
  }
  return out;
}

function castValue(s: string, typ?: string): any {
  if (!typ) return s;
  const t = typ.toLowerCase();
  if (s === '' && t !== 'str') return null;
  try {
    if (t === 'int' || t === 'integer') return parseInt(s, 10);
    if (t === 'float' || t === 'double' || t === 'number') return parseFloat(s);
    if (t === 'bool' || t === 'boolean') {
      const ls = s.toLowerCase();
      if (['true', '1', 'yes', 'y'].includes(ls)) return true;
      if (['false', '0', 'no', 'n'].includes(ls)) return false;
      return Boolean(s);
    }
    return s;
  } catch (e) {
    return s;
  }
}

export function chunkToon(objects: JSObject[], chunkSize = 100, fields?: string[], delimiter = '|', types?: string[], includeSchema = false) {
  if (chunkSize <= 0) throw new Error('chunkSize must be > 0');
  const total = Math.ceil(objects.length / chunkSize);
  const chunks: any[] = [];
  for (let i = 0; i < total; i++) {
    const start = i * chunkSize;
    const end = Math.min(start + chunkSize, objects.length);
    const subset = objects.slice(start, end);
    const toon = toToon(subset, fields, delimiter, types, includeSchema && i === 0);
    chunks.push({ chunkIndex: i, total, start, end, toon });
  }
  return chunks;
}

export function estimateTokens(text: string): number {
  if (!text) return 0;
  const wc = text.split(/\s+/).filter(Boolean).length;
  return Math.floor(wc * 1.3) + 1;
}

export function buildPromptForLLM(instruction: string, objects: JSObject[], fields?: string[], delimiter = '|', types?: string[], includeSchema = false) {
  const toon = toToon(objects, fields, delimiter, types, includeSchema);
  const prompt = `Data (TOON):\n${toon}\n\nInstructions:\n${instruction}`;
  return { prompt, tokensEstimate: estimateTokens(prompt) };
}

File: sdk/toon.py
"""
TOON (Token-Oriented Object Notation) utilities

Provides:
- to_toon(objects, fields=None, delimiter='|') -> str
- from_toon(toon_str, delimiter='|') -> list[dict]
- estimate_tokens(text, model=None) -> int (uses tiktoken when available)
- compare_toon_json_tokens(objects, fields=None, model=None) -> dict
- prompt_with_toon(prompt, objects, fields=None, delimiter='|') -> str

Design choices:
- Delimiter default is '|' (pipe) for readability. Values are escaped for
  backslash, newline, and delimiter.
- Fields order is taken from first object when not provided.
- Token estimation uses tiktoken if installed; otherwise falls back to a
  conservative whitespace-based estimate.

Usage example in toon_demo.py
"""

from __future__ import annotations

import json

# Optional: try to import tiktoken for accurate token counts. If unavailable,
# fall back to heuristic.
try:
    import tiktoken  # type: ignore
except Exception:  # pragma: no cover - optional dependency
    tiktoken = None  # type: ignore


def _escape_value(val: object, delimiter: str) -> str:
    s = "" if val is None else str(val)
    # Escape backslash first
    s = s.replace("\\", "\\\\")
    # Escape newline
    s = s.replace("\n", "\\n")
    # Escape delimiter
    if delimiter:
        s = s.replace(delimiter, f"\\{delimiter}")
    return s


def _unescape_value(s: str, delimiter: str) -> str:
    # Reverse of _escape_value
    res_chars = []
    i = 0
    while i < len(s):
        c = s[i]
        if c == "\\":
            i += 1
            if i >= len(s):
                # trailing backslash, keep it
                res_chars.append("\\")
                break
            nxt = s[i]
            if nxt == "n":
                res_chars.append("\n")
            else:
                # could be escaped delimiter or backslash or other char
                res_chars.append(nxt)
            i += 1
        else:
            res_chars.append(c)
            i += 1
    return "".join(res_chars)


def _split_escaped(line: str, delimiter: str) -> list[str]:
    parts: list[str] = []
    cur_chars: list[str] = []
    i = 0
    while i < len(line):
        c = line[i]
        if c == "\\":
            # include escape and next char as literal sequence to be unescaped later
            if i + 1 < len(line):
                cur_chars.append("\\")
                cur_chars.append(line[i + 1])
                i += 2
            else:
                cur_chars.append("\\")
                i += 1
        elif delimiter and c == delimiter:
            parts.append("".join(cur_chars))
            cur_chars = []
            i += 1
        else:
            cur_chars.append(c)
            i += 1
    parts.append("".join(cur_chars))
    return parts


def to_toon(objects: list[dict], fields: list[str] | None = None, delimiter: str = "|") -> str:
    """Serialize a list of dict-like objects to TOON format.

    - If fields is not provided, fields are inferred from the first object in order of keys.
    - Values are escaped for backslash, newline, and delimiter.
    """
    if objects is None:
        raise ValueError("objects must be a list")
    if len(objects) == 0:
        return ""

    if fields is None:
        # preserve insertion order (Python 3.7+ dict preserves insertion order)
        fields = list(objects[0].keys())

    header = delimiter.join(fields)
    rows: list[str] = []
    for obj in objects:
        row_values: list[str] = []
        for f in fields:
            val = obj.get(f, "")
            row_values.append(_escape_value(val, delimiter))
        rows.append(delimiter.join(row_values))

    return header + "\n" + "\n".join(rows)


def from_toon(toon_str: str, delimiter: str = "|") -> list[dict]:
    """Parse a TOON string back into list of dicts.

    Returns empty list for empty strings.
    """
    if not toon_str:
        return []
    lines = toon_str.splitlines()
    if len(lines) == 0:
        return []
    header_line = lines[0]
    fields = header_line.split(delimiter) if header_line else []
    objs: list[dict] = []
    for line in lines[1:]:
        parts = _split_escaped(line, delimiter)
        values = [_unescape_value(p, delimiter) for p in parts]
        # If row has fewer parts than fields, pad with empty string
        while len(values) < len(fields):
            values.append("")
        obj = {fields[i]: values[i] for i in range(len(fields))}
        objs.append(obj)
    return objs


def estimate_tokens(text: str, model: str | None = None) -> int:
    """Estimate token count for a string for a given model.

    If tiktoken is available, use it. Otherwise fall back to heuristic (words * 1.3).
    """
    if not text:
        return 0
    if tiktoken is not None:
        try:
            # Map some common model names to tiktoken encodings conservatively
            if model is None:
                enc = tiktoken.encoding_for_model("gpt-4")
            else:
                # If specific model not found this may raise; catch and fallback
                enc = tiktoken.encoding_for_model(model)
            return len(enc.encode(text))
        except Exception:
            # fallback below
            pass
    # conservative heuristic: average 1.3 tokens per word
    word_count = len(text.split())
    return int(word_count * 1.3) + 1


def compare_toon_json_tokens(
    objects: list[dict],
    fields: list[str] | None = None,
    model: str | None = None,
    delimiter: str = "|",
) -> dict:
    """Return comparison of token counts between JSON and TOON.

    Returns dict with keys: json_text, toon_text, json_tokens, toon_tokens,
    savings_tokens, savings_percent
    """
    json_text = json.dumps(objects, ensure_ascii=False)
    toon_text = to_toon(objects, fields=fields, delimiter=delimiter)
    json_tokens = estimate_tokens(json_text, model=model)
    toon_tokens = estimate_tokens(toon_text, model=model)
    savings = json_tokens - toon_tokens
    pct = (savings / json_tokens * 100) if json_tokens > 0 else 0.0
    return {
        "json_text": json_text,
        "toon_text": toon_text,
        "json_tokens": json_tokens,
        "toon_tokens": toon_tokens,
        "savings_tokens": savings,
        "savings_percent": round(pct, 2),
    }


def prompt_with_toon(
    prompt: str, objects: list[dict], fields: list[str] | None = None, delimiter: str = "|"
) -> str:
    """Prepend TOON-formatted data to a prompt, separated by two newlines.

    Useful helper when sending data + instructions to an LLM. Keeps the format
    compact while leaving instructions readable.
    """
    toon_text = to_toon(objects, fields=fields, delimiter=delimiter)
    if toon_text:
        return f"Data (TOON):\n{toon_text}\n\nInstructions:\n{prompt}"
    else:
        return prompt


if __name__ == "__main__":
    # Quick smoke demo when run directly
    sample = [
        {"id": 1, "name": "Widget A", "price": 9.99, "description": "Small widget\n2 colors"},
        {"id": 2, "name": "Widget B", "price": 15.5, "description": "Large | heavy"},
    ]
    cmp = compare_toon_json_tokens(sample, model=None)
    print("JSON tokens approx:", cmp["json_tokens"])
    print("TOON tokens approx:", cmp["toon_tokens"])
    print("Savings:", cmp["savings_tokens"], f"({cmp['savings_percent']}%)")
    print("\nTOON representation:\n")
    print(cmp["toon_text"])
